kong/api/arguments.lua

```diff
@@ -279,6 +281,63 @@ infer = function(args, schema)
 end
 
 
+local function decode_map_arg(name, value, container)
+  -- the meaning of square brackets varies depending on the method.
+  -- It is considered a map if the method is POST, PUT, or PATCH
+  -- otherwise square brackets are interpreted as LHS (EE only)
+  local method = get_method()
+  if method ~= "POST" and method ~= "PUT" and method ~= "PATCH" then
+    return nil, "maps not supported for this method"
+  end
+
+  container = container or {}
+
+  local keys = {}
+  local search = name
+
+  while true do
```

Comment:
I started splitting things and soon realized that arrays and maps are pretty much the same thing on a high level, so got rid of the two separate functions that handled maps and arrays separately and split out all the rest to handle individual tasks:

- decode_complex_arg --> decodes a map, an array, a mixed map/array (yes, this is new) or a simple key=value pair, this was needed to keep the outer facing interface unchanged
- decode_map_array_arg --> decodes maps, arrays, or mixed structures
- get_param_name_and_keys (what you said)
- nest_path (what you said)
- extract_param_keys --> your "split_keys_in_bracket", this last one was tricky because we also support things like: `name[foo].test[bar][21].key`